docs: add OAuth 2.0 user doc (#761)
bytebase/bytebase.com: content/docs/administration/sso/oauth2.md

---
title: OAuth 2.0
---

> OAuth (Open Authorization) is an open standard for token-based authentication and authorization which is used to provide single sign-on (SSO). Bytebase supports to create popular OAuth 2.0 providers(GitHub, Google, GitLab) and custom providers.

Reviewer: ```suggestion
OAuth is an open standard for token-based authentication and authorization which is used to provide Single Sign-On (SSO). Bytebase supports popular OAuth 2.0 providers (GitHub, Google, GitLab) and custom providers.
```

![create-sso-dialog](/static/docs/administration/sso/create-sso-dialog.webp)

> In the creating SSO dialog, you need to fill in most of required fields. Including:

Reviewer: ```suggestion
In the creating SSO dialog, you need to fill following fields:
```

> - Basic information:

Reviewer: I feel for the entire section should use a ## Overview

and each should use ### Basic information

  - **Name** is the display name of your SSO item;

>   - **ResourceID** should be an unique string about this resource;

Reviewer: ```suggestion
  - **Resource ID** should be a unique string identifying this resource by Bytebase;
```

  - **Domain** is a SSO domain name;

- Identity provider information: The information is the base concept of [OAuth 2.0](https://oauth.net/2/) and comes from your provider.

  - **Client ID** is a public identifier of the custom provider;
  - **Client Secret** is the OAuth2 client secret from identity provider;
  - **Auth URL** is the custom provider's OAuth2 login page address;

>   - **Scopes** is the scope parameter carried when accessing the OAuth2 URL, which is filled in according to the requirements of the custom provider;

Reviewer: ```suggestion
  - **Scopes** is the scope parameter carried when accessing the OAuth2 URL, which is filled in according to the custom provider;
```

>   - **Token URL** is the API address for obtaining accessToken;

Reviewer: Should accessToken be access token?

>   - **User information URL** is the API address for obtaining user information by accessToken;

Reviewer: These fields are not shown in the above screenshot

> - User information field mapping:

Reviewer: No screenshot

  For different providers, the structure returned by their user information API usually not the same. The information we need to know about an OAuth2 service is how to map the user information into Bytebase user fields. To extract the information we need from the different OAuth2 provider user information API, you need to fill the user information fields mapping form.

  Bytebase will use this information to import the user profile fields when creating new accounts.
  The most important user field mapping is the identifier which are used to identify the Bytebase account associated with the OAuth 2.0 login.

  - **Identifier** is the field name of the unique identifier in 3rd-party user info;
  - **DisplayName** is the field name of display name in 3rd-party user info;
  - **Email** is the field name of primary email in 3rd-party user info.

<hint-block type="info">

OAuth 2.0 requires an authorization callback url in the configuration. Please make sure the [--external-url](/docs/get-started/install/external-url) is set correctly. See more in [Configure External URL](/docs/get-started/install/external-url).

</hint-block>

> ## Create OAuth 2.0 provider with templates

Reviewer: Built-in OAuth Provider

![CleanShot 2023-02-01 at 12 33 27](https://user-images.githubusercontent.com/230323/215949520-edb97265-71a0-461d-9893-5ad85dda3512.png)

Long title should be avoided

> Bytebase provides some popular OAuth 2.0 provider templates including GitHub, GitLab and Google.

Reviewer: ```suggestion
Bytebase provides templates for configuring built-in OAuth providers.
```

### GitHub

1. Follow the [Creating an OAuth App in GitHub](https://docs.github.com/en/developers/apps/building-oauth-apps/creating-an-oauth-app) to create an OAuth app in GitHub.

   ![github-oauth-app-config](/static/docs/administration/sso/github-oauth-app-config.webp)

   <hint-block type="info">

>    The Authorization callback URL is related with Bytebase's external URL.

Reviewer: Need to provide link to the external URL doc. https://bytebase-com-git-steven-byt-2356-oauth2-user-doc-bytebase.vercel.app/docs/get-started/install/external-url

Also, instead of using related, be more explicit about the rules.

   </hint-block>

2. Open creating SSO dialog and select the **GitHub** template.

   ![oauth2-github](/static/docs/administration/sso/oauth2-github.webp)

3. Update the **Client ID** and **Client secret** fields with the GitHub OAuth App you just created.
4. After filling in all the required fields, try to click **Test connection**.
5. If everything is OK, click the **Create** button.

### GitLab

1. Follow the [Configure GitLab as an OAuth 2.0 authentication identity provider](https://docs.github.com/en/developers/apps/building-oauth-apps/creating-an-oauth-app) to create an OAuth 2 application in GitLab.

   ![gitlab-oauth-app-config](/static/docs/administration/sso/gitlab-oauth-app-config.webp)

> 2. Open creating SSO dialog and select the **GitLab** template.

Reviewer: If user just jumps to the GitLab section, they don't know what to fill for Redirect URI

Author: <img width="869" alt="image" src="https://user-images.githubusercontent.com/24653555/215998155-6cec68a5-8cad-4bd8-84fd-7227856223ab.png">

Added a hint block in the beginning of section.

---
title: OAuth 2.0
---

> OAuth is an open standard for token-based authentication and authorization which is used to provide single sign-on (SSO). Bytebase supports to create popular OAuth 2.0 providers(GitHub, Google, GitLab) and custom providers.

Reviewer: ```suggestion
OAuth is an open standard for token-based authentication and authorization which is used to provide Single Sign-On (SSO). Bytebase supports to create popular OAuth 2.0 providers (GitHub, Google, GitLab) and custom providers.
```
